fix(frontend): make `path_to` breadth-first

W95Psp opened the pull request:
Fixes #692

Status: seems like this is introducing a bug, I have to debug.
Some implementations cannot be found after the change

maximebuyse commented:
> A first fix to this implementation (or to a bug revealed by it) is in #797
> 
> There is another problem remaining (reaching https://github.com/hacspec/hax/blob/main/frontend/exporter/src/traits.rs#L349)

This bug introduced by the initial version of this PR is now fixed. It was just that switching from a recursive depth-first search to a breadth-first search with a queue meant the parents and associated traits had to be taken from the candidate path and not `self`. In other words we were only checking paths of length 1 with `path_to`.

W95Psp reviewed (commented):
Nice, thank you @maximebuyse! That looks good to me, and @franziskuskiefer is currently trying to break it by throwing it on bigger crates :sweat_smile: :crossed_fingers:

franziskuskiefer commented:
The frontend panics now in most cases where it looped before.

```
thread 'rustc' panicked at /rustc/6b0f4b5ec3aa707ecaa78230722117324a4ce23c/compiler/rustc_type_ir/src/binder.rs:713:9:
type parameter `T/#1` (T/#1/1) out of range when instantiating, args=[Alias(Projection, AliasTy { args: [Self/#0], def_id: DefId(5:71 ~ crypto_common[e66d]::BlockSizeUser::BlockSize) })]
```

W95Psp commented:
This is #683, which is not related. IMO we should merge that PR, that's fixing the looping issue

franziskuskiefer commented:
> that's fixing the looping issue

How do we know that it is fixing the issue?

maximebuyse commented:
> > that's fixing the looping issue
> 
> How do we know that it is fixing the issue?

You're right, it doesn't work on #692 . The looping behavior was gone because the breadth-first search had a bug.

W95Psp commented:
Oh, you mean that it's still looping? :/

W95Psp commented:
I pushed `make-instantiating-bug-nonfatal` which is making the instantiating bug non-fatal. The looping behavior is gone on https://github.com/RustCrypto/traits/tree/master/cipher at least.

W95Psp commented on frontend/exporter/src/traits.rs:
```diff
@@ -200,62 +199,90 @@ pub mod rustc {
-                    return Some(vec![]);
+        #[tracing::instrument(level = "trace", skip(s, param_env))]
+        pub fn path_to<'tcx, S: UnderOwnerState<'tcx>>(
+            starting_points: Vec<AnnotatedPredicate<'tcx>>,
```
Maybe you can take a `&[AnnotatedPredicate<'tcx>]` instead of a vec here?
Then you would not need to clone on call site.

W95Psp commented on frontend/exporter/src/traits.rs:
```diff
@@ -200,62 +199,90 @@ pub mod rustc {
+                        origin: pred.clone(),
+                    })
+                })
+                .filter(|candidate| candidate.is_some())
+                .map(|candidate| candidate.unwrap())
```
Use one [`filter_map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map) instead of map+filter+map here, we'll get rid of the unwrap :) in general we don't want `unwrap`s at all. Btw, you can use `option.s_unwrap(the_state_of_type_State)` if you really ever need to unwrap: that unwraps with a spanned message.

W95Psp commented on frontend/exporter/src/traits.rs:
```diff
@@ -200,62 +199,90 @@ pub mod rustc {
+                    candidates.push_back(Candidate {
+                        pred: parent_pred.clone(),
+                        path,
+                        origin: candidate.origin.clone(),
```
I think we can make `AnnotatedPredicate` copy. Can you add `Copy` to the derives of `rustc_utils::AnnotatedPredicate`, and then remove `clone`s?

W95Psp reviewed (commented):
LGTM, with a few nits!

maximebuyse commented:
> LGTM, with a few nits!

Thanks for the tips, I made the corrections.

W95Psp commented:
Nice, thanks! Have you tried the patch on both https://github.com/hacspec/hax/issues/686 and https://github.com/hacspec/hax/issues/692? (I updated `make-instantiating-bug-nonfatal`)
If #692 goes through with `make-instantiating-bug-nonfatal`, let's document that and close it in favor of https://github.com/hacspec/hax/issues/747.

Also, you can resolve conversations on the comments I added, otherwise the PR will be blocked from merging :)

maximebuyse commented:
> Nice, thanks! Have you tried the patch on both #686 and #692? (I updated `make-instantiating-bug-nonfatal`) If #692 goes through with `make-instantiating-bug-nonfatal`, let's document that and close it in favor of #747.
> 
> Also, you can resolve conversations on the comments I added, otherwise the PR will be blocked from merging :)

This seems to solve #692, I added a test with this example.
But for #686, no change (the memory consumption keeps increasing until I have to kill it). So it might be another problem. 
I didn't try with `make-instantiating-bug-non-fatal` but I didn't get instantiating issues apparently.

franziskuskiefer reviewed (approved):
Alright, then let's get this in and keep looking for the issue in #686.

W95Psp commented:
Mmh, ok. Added a comment there: https://github.com/hacspec/hax/issues/686#issuecomment-2255337646.
